Connection Fails When There is a Special Character in SMB Password

**Description:**
<!-- Give a brief description of the issue -->
Observing the bellow error while connecting to SMB2 when there are special characters in the password.
`Caused by: org.apache.commons.vfs2.FileSystemException: Badly formed URI "smb2://d;testuser1:$1$2$3\$\\$@192.168.74.123/secured/tmp/ziel_sz2sz_1".
	at org.apache.commons.vfs2.provider.smb2.Smb2FileNameParser.parseURIString(Smb2FileNameParser.java:109)
	at org.apache.commons.vfs2.provider.smb2.Smb2FileNameParser.parseUri(Smb2FileNameParser.java:69)
	at org.apache.commons.vfs2.provider.smb2.Smb2FileProvider.parseUri(Smb2FileProvider.java:71)
	at org.apache.commons.vfs2.provider.AbstractOriginatingFileProvider.findFile(AbstractOriginatingFileProvider.java:75)
	... 20 more`

**Affected Product Version:**
EI 6.6.0

**Steps to reproduce:**

- setup a SMB server locally.
- Use password as **$1$\\%40%2f$4\$5**.
- Try to connect to the SMB2 using a VFS proxy.

--- a/commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2FileNameParser.java
+++ b/commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2FileNameParser.java
@@ -25,6 +25,9 @@ import org.apache.commons.vfs2.provider.UriParser;
 import org.apache.commons.vfs2.provider.VfsComponentContext;
 
 import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 public class Smb2FileNameParser extends HostFileNameParser {
 
@@ -34,9 +37,25 @@ public class Smb2FileNameParser extends HostFileNameParser {
 
     private static final char[] RESERVED_CHARS = { '#' };
 
+    List<Character> validPasswordChars;
+
     protected Smb2FileNameParser() {
 
         super(PORT);
+
+        // rfc3986
+        validPasswordChars = new ArrayList<Character>(
+                Arrays.asList('/', '-', '.', '_', '~', '%', '!', '$', '&', '\'', '(', ')',
+                        '*', '+', ',', ';', '=', ':', '@'));
+        for (char c = 'a'; c <= 'z'; c++) {
+            validPasswordChars.add(c);
+        }
+        for (char c = 'A'; c <= 'Z'; c++) {
+            validPasswordChars.add(c);
+        }
+        for (char c = '0'; c <= '9'; c++) {
+            validPasswordChars.add(c);
+        }
     }
 
     public static Smb2FileNameParser getInstance() {
@@ -46,7 +65,7 @@ public class Smb2FileNameParser extends HostFileNameParser {
 
     @Override
     public boolean encodeCharacter(final char ch) {
-        return super.encodeCharacter(ch) || ch == '?' || ch == ' ' || ch == '#';
+        return super.encodeCharacter(ch) || !validPasswordChars.contains(ch);
     }
 
     /**

--- a/commons-vfs2-sandbox/src/test/java/org/apache/commons/vfs2/provider/smb2/test/Smb2FileNameParserTest.java
+++ b/commons-vfs2-sandbox/src/test/java/org/apache/commons/vfs2/provider/smb2/test/Smb2FileNameParserTest.java
@@ -32,7 +32,7 @@ public class Smb2FileNameParserTest {
     private final static String testPasswordUE = /* printable ascii chars */
             "%20%21%22%23%24%25%26%27%28%29%2A%2B%2C%2D%2E%2F%30%31%32%33%34%35%36%37%38%39%3A%3B%3C%3D%3E%3F%40%41%42%43%44%45%46%47%48%49%4A%4B%4C%4D%4E%4F" +
                     "%50%51%52%53%54%55%56%57%58%59%5A%5B%5C%5D%5E%5F%60%61%62%63%64%65%66%67%68%69%6A%6B%6C%6D%6E%6F%70%71%72%73%74%75%76%77%78%79%7A%7B%7C%7D%7E";
-    private final static String testPath = /* printable ascii chars without / and \ */
+    private final static String testPath = 
             "%20%21%22%23%24%25%26%27%28%29%2A%2B%2C%2D%2E%2F%30%31%32%33%34%35%36%37%38%39%3A%3B%3C%3D%3E%3F%40%41%42%43%44%45%46%47%48%49%4A%4B%4C%4D%4E%4F" +
                     "%50%51%52%53%54%55%56%57%58%59%5A%5B%5C%5D%5E%5F%60%61%62%63%64%65%66%67%68%69%6A%6B%6C%6D%6E%6F%70%71%72%73%74%75%76%77%78%79%7A%7B%7C%7D%7E";
     private final static String testURIString = "smb2://WORKGROUP;USER:" + testPasswordUE + "@example.com/ShareName/Folder1/" + testPath + "/Folder2/?somequery=true&another=query#fragment";

--- a/commons-vfs2-sandbox/src/test/java/org/apache/commons/vfs2/provider/smb2/test/Smb2FileNameParserTest.java
+++ b/commons-vfs2-sandbox/src/test/java/org/apache/commons/vfs2/provider/smb2/test/Smb2FileNameParserTest.java
@@ -1,0 +1,59 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.commons.vfs2.provider.smb2.test;
+
+import org.apache.commons.vfs2.FileSystemException;
+import org.apache.commons.vfs2.provider.smb2.Smb2FileName;
+import org.apache.commons.vfs2.provider.smb2.Smb2FileNameParser;
+import org.junit.Test;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
+
+public class Smb2FileNameParserTest {
+    private final static String testPasswordUE = /* printable ascii chars */
+            "%20%21%22%23%24%25%26%27%28%29%2A%2B%2C%2D%2E%2F%30%31%32%33%34%35%36%37%38%39%3A%3B%3C%3D%3E%3F%40%41%42%43%44%45%46%47%48%49%4A%4B%4C%4D%4E%4F" +
+                    "%50%51%52%53%54%55%56%57%58%59%5A%5B%5C%5D%5E%5F%60%61%62%63%64%65%66%67%68%69%6A%6B%6C%6D%6E%6F%70%71%72%73%74%75%76%77%78%79%7A%7B%7C%7D%7E";
+    private final static String testPath = /* printable ascii chars without / and \ */
+            "%20%21%22%23%24%25%26%27%28%29%2A%2B%2C%2D%2E%2F%30%31%32%33%34%35%36%37%38%39%3A%3B%3C%3D%3E%3F%40%41%42%43%44%45%46%47%48%49%4A%4B%4C%4D%4E%4F" +
+                    "%50%51%52%53%54%55%56%57%58%59%5A%5B%5C%5D%5E%5F%60%61%62%63%64%65%66%67%68%69%6A%6B%6C%6D%6E%6F%70%71%72%73%74%75%76%77%78%79%7A%7B%7C%7D%7E";
+    private final static String testURIString = "smb2://WORKGROUP;USER:" + testPasswordUE + "@example.com/ShareName/Folder1/" + testPath + "/Folder2/?somequery=true&another=query#fragment";
+
+    @Test(expected = Test.None.class /* no exception expected */)
+    public void testParseUriUserInfo() throws FileSystemException, URISyntaxException {
+        String expectedDecodedUserInfo = new URI(testURIString).getUserInfo();
+
+        Smb2FileName fileName = (Smb2FileName) Smb2FileNameParser.getInstance().parseUri(null, null, testURIString);
+        String resultDecodedUserInfo = fileName.getUserName() + ":" + fileName.getPassword();
+
+        assertThat("Password decoded correctly", resultDecodedUserInfo, is(expectedDecodedUserInfo));
+    }
+
+    @Test(expected = Test.None.class /* no exception expected */)
+    public void testParseUriPath() throws FileSystemException, URISyntaxException {
+        String expectedDecodedPath = (new URI(testURIString).getPath());
+
+        Smb2FileName fileName = (Smb2FileName) Smb2FileNameParser.getInstance().parseUri(null, null, testURIString);
+        String resultDecodedPath = "/" + fileName.getShareName() + fileName.getPathDecoded() + "/";
+
+        assertThat("Path decoded correctly", resultDecodedPath, is(expectedDecodedPath));
+    }
+}

--- a/commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/GenericFileName.java
+++ b/commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/GenericFileName.java
@@ -24,8 +24,14 @@ import org.apache.commons.vfs2.FileType;
  * password), hostname, port, and path.
  */
 public class GenericFileName extends AbstractFileName {
-    private static final char[] USERNAME_RESERVED = { ':', '@', '/' };
-    private static final char[] PASSWORD_RESERVED = { '@', '/', '?' };
+    private static final char[] USERNAME_RESERVED = {
+            0x20, 0x22, 0x23, 0x2D, 0x2E, 0x2F, 0x3C, 0x3E, 0x3F, 0x40,
+            0x5B, 0x5C, 0x5D, 0x5E, 0x5F, 0x60, 0x7B, 0x7C, 0x7D, 0x7E
+    };
+    private static final char[] PASSWORD_RESERVED = {
+            0x20, 0x22, 0x23, 0x2D, 0x2E, 0x2F, 0x3C, 0x3E, 0x3F, 0x40,
+            0x5B, 0x5C, 0x5D, 0x5E, 0x5F, 0x60, 0x7B, 0x7C, 0x7D, 0x7E
+    };
     private final String userName;
     private final String hostName;
     private final int defaultPort;

